src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java

```diff
@@ -0,0 +1,593 @@
+import org.apache.cassandra.db.rows.UnfilteredRowIterator;
+import org.apache.cassandra.dht.AbstractBounds;
+import org.apache.cassandra.dht.Token;
+import org.apache.cassandra.exceptions.RequestTimeoutException;
+import org.apache.cassandra.index.Index;
+import org.apache.cassandra.index.sai.QueryContext;
+import org.apache.cassandra.index.sai.analyzer.AbstractAnalyzer;
+import org.apache.cassandra.index.sai.metrics.TableQueryMetrics;
+import org.apache.cassandra.index.sai.utils.PrimaryKey;
+import org.apache.cassandra.index.sai.utils.PrimaryKeyFactory;
+import org.apache.cassandra.index.sai.utils.RangeIterator;
+import org.apache.cassandra.io.util.FileUtils;
+import org.apache.cassandra.schema.TableMetadata;
+import org.apache.cassandra.utils.AbstractIterator;
+
+public class StorageAttachedIndexSearcher implements Index.Searcher
+{
+    private final ReadCommand command;
+    private final QueryController controller;
+    private final QueryContext queryContext;
+    private final PrimaryKeyFactory keyFactory;
+
+    public StorageAttachedIndexSearcher(ColumnFamilyStore cfs,
+                                        TableQueryMetrics tableQueryMetrics,
+                                        ReadCommand command,
+                                        RowFilter filterOperation,
+                                        long executionQuotaMs)
+    {
+        this.command = command;
+        this.queryContext = new QueryContext(command, executionQuotaMs);
+        this.controller = new QueryController(cfs, command, filterOperation, queryContext, tableQueryMetrics);
+        this.keyFactory = PrimaryKey.factory(cfs.metadata().comparator);
+    }
+
+    @Override
+    public ReadCommand command()
+    {
+        return command;
+    }
+
+    @Override
+    public PartitionIterator filterReplicaFilteringProtection(PartitionIterator fullResponse)
+    {
+        for (RowFilter.Expression expression : controller.filterOperation())
+        {
+            AbstractAnalyzer analyzer = controller.getContext(expression).getIndexAnalyzerFactory().create();
+            try
+            {
+                if (analyzer.transformValue())
+                    return applyIndexFilter(fullResponse, analyzeFilter(), queryContext);
+            }
+            finally
+            {
+                analyzer.end();
+            }
+        }
+
+        // if no analyzer does transformation
+        return Index.Searcher.super.filterReplicaFilteringProtection(fullResponse);
+    }
+
+    @Override
+    public UnfilteredPartitionIterator search(ReadExecutionController executionController) throws RequestTimeoutException
+    {
+        return new ResultRetriever(analyze(), analyzeFilter(), controller, executionController, queryContext, keyFactory);
+    }
+
+    /**
+     * Converts expressions into filter tree for query.
+     *
+     * @return operation
+     */
+    private RangeIterator analyze()
+    {
+        return Operation.buildIterator(controller);
```

Comment:
I think we inherited this naming from SASI. Why don't we just inline the calls to `buildIterator()` and `buildFilter()`, adding any JavaDoc we want to those methods. In the `Searcher` here, the naming something around "analysis" is pulling a random detail from `Operation` that we more or less don't care about.

Reply:
Yes, this makes perfect sense. I have been wanting to simplify this path for a long time. This will make debugging simpler because everything is now processed in the `ResultRetriever` constructor. At the same time I have renamed `controller` to `queryController` to try and get a bit of naming consistency.